Source/WebCore/loader/ResourceMonitor.cpp

```diff
@@ -0,0 +1,185 @@
+ * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "config.h"
+#include "ResourceMonitor.h"
+
+#include "LocalFrame.h"
+#include <wtf/CryptographicallyRandomNumber.h>
+
+namespace WebCore {
+
+#if ENABLE(CONTENT_EXTENSIONS)
+
+static constexpr size_t sNetworkUsageThreshold = 4 * 1024 * 1024; // 4MB
+static constexpr size_t sNetworkUsageThresholdRandomness = 1303 * 1024; // 1303KB
+
+static size_t networkUsageThresholdWithRandomNoise()
+{
+    return sNetworkUsageThreshold + static_cast<size_t>(cryptographicallyRandomUnitInterval() * sNetworkUsageThresholdRandomness);
+}
+
+ResourceMonitor::ResourceMonitor(LocalFrame& frame)
+    : m_frame(frame)
+    , m_networkUsageThreshold { networkUsageThresholdWithRandomNoise() }
+{
+    if (RefPtr parentMonitor = parentResourceMonitor())
+        setEligibility(parentMonitor->eligibility());
+}
+
+void ResourceMonitor::setEligibility(Eligibility eligibility)
+{
+    if (m_eligibility == eligibility || m_eligibility == Eligibility::Eligible)
+        return;
+
+    m_eligibility = eligibility;
+
+    if (RefPtr parentMonitor = parentResourceMonitor())
+        parentMonitor->setEligibility(eligibility);
+    else
+        checkCondition();
+}
+
+void ResourceMonitor::setDocument(URL&& url)
+{
+    m_frameURL = WTFMove(url);
+    didReceiveResource(m_frameURL, ContentExtensions::ResourceType::Document);
+}
+
+void ResourceMonitor::didReceiveResource(const URL& url, ResourceTypeSet resourceType)
+{
+    if (m_eligibility == Eligibility::Eligible)
+        return;
+
+    RefPtr page = protectedFrame()->mainFrame().page();
+    if (!page)
+        return;
+
+    ResourceLoadInfo info = {
+        .resourceURL = url,
+        .mainDocumentURL = page->mainFrameURL(),
+        .frameURL = m_frameURL,
+        .type = resourceType
+    };
+
+    Checker::singleton().checkEligibility(info, [weakThis = WeakPtr(this)](Eligibility eligibility) mutable {
+        callOnMainRunLoop([weakThis = WTFMove(weakThis), eligibility] {
+            if (RefPtr protectedThis = weakThis.get())
+                protectedThis->setEligibility(eligibility);
+        });
+    });
+}
+
+void ResourceMonitor::addNetworkUsage(size_t bytes)
+{
+    m_networkUsage += bytes;
+
+    if (RefPtr parentMonitor = parentResourceMonitor())
+        parentMonitor->addNetworkUsage(bytes);
+    else
+        checkCondition();
+}
+
+void ResourceMonitor::checkCondition()
```

Comment:
`checkCondition()` is too generic a name. Maybe `checkNetworkUsageExcessIfNecessary`